scipy/stats/_qmc.py

```diff
@@ -898,6 +988,151 @@ def random(self, n: IntNumber = 1) -> np.ndarray:
+
+    References
+    ----------
+    .. [1] Jin et al. "An efficient algorithm for constructing optimal design
+       of computer experiments", Journal of Statistical Planning and
+       Inference, 2005.
+    .. [2] Damblin et al., "Numerical studies of space filling designs:
+       optimization of Latin Hypercube Samples and subprojection properties."
+       Journal of Simulation, 2013.
+
+    Examples
+    --------
+    Generate samples from an optimal design.
+
+    >>> from scipy.stats import qmc
+    >>> sampler = qmc.OptimalLatinHypercube(d=2)
+    >>> sample = sampler.random(n=5)
+    >>> sample
+    array([[0.84052691, 0.53664833],  # random
+           [0.1545328 , 0.36265316],
+           [0.52177809, 0.93343721],
+           [0.68033825, 0.06474907],
+           [0.26544879, 0.61163943]])
+
+    Compute the quality of the sample using the discrepancy criterion.
+
+    >>> qmc.discrepancy(sample)
+    0.01602810655510689  # random
+
+    You can possibly improve the quality of the sample by performing more
+    optimization iterations by using `niter`:
+
+    >>> sampler_2 = qmc.OptimalLatinHypercube(d=2, n_perturbations=5000)
+    >>> sample_2 = sampler_2.random(n=5)
+    >>> qmc.discrepancy(sample_2)
+    0.01600398742881648  # random
+
+    """
+
+    def __init__(self, d: IntNumber, *, start_sample: npt.ArrayLike = None,
+                 n_perturbations: IntNumber = 10000,
+                 seed: SeedType = None):
+        super().__init__(d=d, seed=seed)
+        self.n_perturbations = n_perturbations
+        self.start_sample: Optional[np.ndarray] = None
+
+        if start_sample is not None:
+            self.start_sample = np.asarray(start_sample).copy()
+            self.start_disc = discrepancy(self.start_sample)
+        else:
+            self.start_disc = np.inf
+
+        self.lhs = LatinHypercube(self.d, seed=self.rng)
+
+    def random(self, n: IntNumber = 1):
+        """Draw `n` in the half-open interval ``[0, 1)``.
+
+        Parameters
+        ----------
+        n : int, optional
+            Number of samples to generate in the parameter space. Default is 1.
+
+        Returns
+        -------
+        sample : array_like (n, d)
+            Optimal LHS sample.
+
+        """
+        if self.d == 0 or n == 0:
+            return np.empty((n, self.d))
+
+        if self.start_sample is None:
+            best_sample = self.lhs.random(n)
+            best_disc = discrepancy(best_sample)
+        else:
+            best_sample = self.start_sample
+            best_disc = self.start_disc
+
+        if n == 1:
+            return best_sample
+
+        bounds = ([0, self.d - 1],
+                  [0, n - 1],
+                  [0, n - 1])
+
+        for _ in range(self.n_perturbations):
+            col = rng_integers(self.rng, *bounds[0])
+            row_1 = rng_integers(self.rng, *bounds[1])
+            row_2 = rng_integers(self.rng, *bounds[2])
+            doe = best_sample.copy()
```

Comment:
If `_perturb_discrepancy` just predicts the new discrepancy (without mutating its first argument), is this needed? A copy of `start_disc` should be made once at the beginning, but not every iteration. 

(If `_perturb_discrepancy` _does_ mutate the array, it should probably return the mutated array, or it should perform the copy itself.)

Reply:
Totally! Maybe a leftover of the previous version...